### PR TITLE
WIP Add InlineVolumeSource to VolumeAttachment API for supporting Inline Volume migration

### PR DESCRIPTION
**Pull request by davidz627**
Design here: https://github.com/kubernetes/community/pull/3291

Adding a generic `VolumeSource` to the `VolumeAttachment` API so that we can shim in-tree in-line volumes to the CSI Attacher to be serviced by the CSI backend.

/kind api-change
/kind feature
/sig storage
/assign @saad-ali @msau42 @thockin 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/api/storage/v1beta1/types.go [hunk 1]
```diff
@@ -166,7 +166,12 @@ type VolumeAttachmentSource struct {
 	// +optional
 	PersistentVolumeName *string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
 
-	// Placeholder for *VolumeSource to accommodate inline volumes in pods.
+	// Copy of VolumeSource to accommodate VolumeSource from a pod
+	// to support shimming of in-tree inline volumes to a CSI backend.
+	// This field is alpha-level and is only honored by servers that
+	// enable the CSIMigration feature.
+	// +optional
+	InlineVolumeSource *v1.VolumeSource `json:"inlineVolumeSource,omitempty" protobuf:"bytes,2,opt,name=inlineVolumeSource"`
```
Comment (thockin) [inline staging/src/k8s.io/api/storage/v1beta1/types.go]: Could this plausibly be a CSI spec inside here?

**Review by thockin — COMMENTED** [review]
API changes LGTM.  Ping me when the rest is reviewed.

File: staging/src/k8s.io/api/storage/v1beta1/types.go (hunk 1 above)
Comment (davidz627) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: Not exactly sure which question you're asking so I will answer it the two ways I understood:
We are not adding the CSI Spec inside of the `VolumeSource` struct so we are not leaking CSI VolumeSource into this field.
The `CSIVolumeSource` (https://github.com/kubernetes/kubernetes/pull/74086/commits/5b64d68c5141de77acbda0de94b0a138fe6ce44f#diff-4143fe2cdf42f1748648a9b80aa2ba07R1610) is stripped down to only support `mount/unmount` fields so is not robust enough for all the information the external-attacher needs for shimmed volumes. Therefore we will just send the un-translated in-tree volume source to the external-attacher and translate it on the external-attacher side.

Hopefully that answered your question.

Comment (liggitt) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: could it be a `*CSIPersistentVolumeSource`?

Comment (thockin) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: I guess I was just saying it's sort of recursive.  Not really a problem.

File: pkg/apis/storage/validation/validation.go [hunk 2]
```diff
@@ -165,8 +165,12 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
...
 	allErrs := field.ErrorList{}
-	if source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath, ""))
+	pvNameNotExist := source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0
+	if source.InlineVolumeSource == nil && pvNameNotExist {
+		allErrs = append(allErrs, field.Required(fldPath, "must specify exactly one of InlineVolumeSource and PersistentVolumeName"))
+	}
+	if source.InlineVolumeSource != nil && !pvNameNotExist {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "must specify exactly one of InlineVolumeSource and PersistentVolumeName"))
 	}
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: do we limit to particular volume sources, or allow all? should we call `validateVolumeSource()` on the volume source?

File: pkg/apis/storage/types.go [hunk 3]
```diff
@@ -157,7 +157,12 @@ type VolumeAttachmentSource struct {
 	// +optional
 	PersistentVolumeName *string
 
-	// Placeholder for *VolumeSource to accommodate inline volumes in pods.
+	// Copy of VolumeSource to accommodate VolumeSource from a pod
+	// to support shimming of in-tree inline volumes to a CSI backend.
+	// This field is alpha-level and is only honored by servers that
+	// enable the CSIMigration feature.
+	// +optional
+	InlineVolumeSource *api.VolumeSource
```
Comment (liggitt) [inline pkg/apis/storage/types.go]: some volume sources contain object references which need to refer to things in the source pod's namespace (e.g. ISCSI.SecretRef, Glusterfs.EndpointsName, RBD.SecretRef, Cinder.SecretRef, etc). how is the originating namespace tracked so those can be resolved correctly?

File: pkg/registry/storage/volumeattachment/strategy.go [hunk 4]
```diff
@@ -98,15 +105,21 @@ func (volumeAttachmentStrategy) PrepareForUpdate(ctx context.Context, obj, old r
...
 	case storageapiv1beta1.SchemeGroupVersion:
 		// allow modification of Status via main resource for v1beta1
 	default:
-		newVolumeAttachment := obj.(*storage.VolumeAttachment)
-		oldVolumeAttachment := old.(*storage.VolumeAttachment)
 		newVolumeAttachment.Status = oldVolumeAttachment.Status
 		// No need to increment Generation because we don't allow updates to spec
 	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) && oldVolumeAttachment.Spec.Source.InlineVolumeSource == nil {
```
Comment (liggitt) [inline pkg/registry/storage/volumeattachment/strategy.go]: +1

File: staging/src/k8s.io/api/storage/v1beta1/types.go (hunk 1 above)
Comment (msau42) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: We can explicitly disallow `CSIVolumeSource` in here.  `CSIPersistentVolumeSource` is only part of `PersistentVolumeSource`

Comment (liggitt) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: > `CSIPersistentVolumeSource` is only part of `PersistentVolumeSource`

I know, but the information produced by the inline volume -> csi translator is all present in the CSIPersistentVolumeSource type: namespaced secret references, full volume options, etc

I was asking if it would make sense to use `*CSIPersistentVolumeSource` in this field *instead of* `*VolumeSource`

Comment (msau42) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: Oh that is very intriguing. Will let @davidz627 evaluate.

Comment (davidz627) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: Really interesting idea, and I see how this would solve: https://github.com/kubernetes/kubernetes/pull/74629#discussion_r260589501 as well...

Then we can restrict the allowed volume sources based on feature flags in validation to resolve https://github.com/kubernetes/kubernetes/pull/74629#discussion_r260589421 (as well as always restrict the `CSIPersistentVolumeSource`). I like this idea. Will implement

Comment (davidz627) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: misread comment. I see `CSIPersistentVolumeSource` now. I believe that was the original design but we lost our way in the middle there.

It makes the most sense because it contains all the namespace information. As well as being the actual end artifact required for the external-attacher to do its thing. It makes more sense to do the translation of the in-line volume on the Kubernetes side where we have more of the Pod information than trying to do it on the external attacher side. This also simplifies the API validation.

**Comment by davidz627** [comment]
API part is stable for now. I am working on the implementation in a separate commit (same PR)

File: pkg/apis/storage/types.go (hunk 3 above)
Comment (thockin) [reply pkg/apis/storage/types.go]: ooh, great point

Comment (thockin) [reply pkg/apis/storage/types.go]: Maybe this should be a Pod reference with namespace and name plus a volume name within that pod?  There's security implications there, I guess.  Maybe simply copying namespace here is enough.

**Comment by thockin** [comment]
@davidz627 there are a couple comments open still, but github makes them hard to see in the code tab.  Specifically the point about namespaces.

File: pkg/apis/storage/types.go (hunk 3 above)
Comment (davidz627) [reply pkg/apis/storage/types.go]: This has been resolved by making the `InlineVolumeSource` a `CSIPersistentVolumeSource`. Since `CSIPersistentVolumeSource` is the "end result" that the external-attacher uses to attach the volumes it contains [and must always contain] all the requisite information for attach including secrets and secret namespaces. That then pushes the onus of translating the original inline volume to a `CSIPersistentVolumeSource` to the AttachDetachController which should have all the required information to populate the new field.

File: pkg/apis/storage/validation/validation.go (hunk 2 above)
Comment (davidz627) [reply pkg/apis/storage/validation/validation.go]: we changed to `CSIPersistentVolumeSource`, validating that it is a valid `CSIPersistentVolumeSource`.

File: pkg/apis/storage/types.go [hunk 5]
```diff
@@ -157,7 +157,12 @@ type VolumeAttachmentSource struct {
 	// +optional
 	PersistentVolumeName *string
 
-	// Placeholder for *VolumeSource to accommodate inline volumes in pods.
+	// Translated VolumeSource from a pod to a CSIPersistentVolumeSource
+	// to support shimming of in-tree inline volumes to a CSI backend.
+	// This field is alpha-level and is only honored by servers that
+	// enable the CSIMigration feature.
+	// +optional
+	InlineVolumeSource *api.CSIPersistentVolumeSource
```
Comment (thockin) [inline pkg/apis/storage/types.go]: Thinking forward:

Will we *ever* have a new storage spec after CSI?  Seems unlikely, but possible.  Will that have inline volumes?  Probably.  How could those be expressed here?  We have a CSI-specific field with a generic name.

Comment (liggitt) [reply pkg/apis/storage/types.go]: agree. something like `InlineCSIVolumeSource` leaves us wiggle room (and is more descriptive of the current field)

File: pkg/apis/storage/validation/validation.go [hunk 6]
```diff
@@ -165,8 +166,14 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
 // validateSource tests if the source is valid for VolumeAttachment.
 func validateVolumeAttachmentSource(source *storage.VolumeAttachmentSource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath, ""))
+	pvNameNotExist := source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0
+	if source.InlineVolumeSource == nil && pvNameNotExist {
+		allErrs = append(allErrs, field.Required(fldPath, "must specify exactly one of InlineVolumeSource and PersistentVolumeName"))
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: use json field names in error messages

File: pkg/apis/storage/validation/validation.go [hunk 7]
```diff
@@ -165,8 +166,14 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
...
+	}
+	if source.InlineVolumeSource != nil && !pvNameNotExist {
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: nit: the double-negative is hard to follow... suggest separating the mutually-exclusive validation checks from the validation of InlineVolumeSource:

```go
switch {
case source.InlineVolumeSource == nil && source.PersistentVolumeName == nil:
  ... error about exactly one being required

case source.InlineVolumeSource != nil && source.PersistentVolumeName != nil:
  ... error about exactly one being required

case source.PersistentVolumeName != nil:
  if len(*source.PersistentVolumeName) == 0 {
    ... invalid error
  }

case source.InlineVolumeSource != nil:
  ... call storagevalidation.ValidateCSIPersistentVolumeSource
}
```

**Review by liggitt — COMMENTED** [review]
a couple nits on validation, and s/InlineVolumeSource/InlineCSIVolumeSource/ (or something similar), then API lgtm

File: pkg/apis/storage/validation/validation.go (hunk 7 above)
Comment (davidz627) [reply pkg/apis/storage/validation/validation.go]: ah. The logic was even wrong originally too, thanks for the suggestion

**Comment by liggitt** [comment]
API LGTM

**Comment by davidz627** [comment]
Closed in favor of: https://github.com/kubernetes/kubernetes/pull/77703
